### ISSUE TITLE
App Level Authentication

### ISSUE THREAD

**Issue by MattW86** [realm/realm-mobile-platform#118]
Hi,
ROS is able to handle multiple apps which is great. So, if a user registers to an app previously and then tries to register to another app (considering the user is not aware of the server level authentication), there will be an error. Is there a workaround for this issue?
Thanks.

**Comment by istx25** [comment]
Hi @MattW86. Would you be able to elaborate more on what you mean? Specifically with "an app previously" and then "tries to register to another app." Wouldn't these apps be on different instances of RMP/ROS? Thanks.

**Comment by MattW86** [comment]
Hi @istx25, I meant to say that if I have 2 different apps, lets say App1 and App2, they both can access to one instance of ROS as ROS can potentially support multiple apps or am I wrong about it?

**Comment by Montrazul** [comment]
You are right about it. The "problem" (it is not really a problem, in realms case it is a feature) is that we get the same error when registration failed for some reasons and when it fails because the user is already registered. So it isnt easy to communicate the kind of the error to the user. If the registration fails you cannot say why it is failing. If we could get an error code which is especially telling us that an error occured because the user is already registered we could tell the user "hey, you are already registered".

Ofcourse we can tell the user "an error occured, maybe you are already registered" but in my opinion this has a bad smell like "they guess what the error is without knowing it".

**Comment by MattW86** [comment]
My suggestion is to add one more parameter to sync credentials 
for example  
`SyncCredentials.usernamePassword(username: "username", password: "password", appname: "myapp")`

**Comment by MattW86** [comment]
@istx25 
I would like to explain why we would like to see this feature. We believe multi tenant approach would be better for individual devs like us. For instance, we developed about 20 apps using Realm in the last 2 years and at least 10 of them do not have any active users. It will be too expensive to run 10 instances of ROS on Amazon for apps that are barely being used. Maintaining servers is also  time consuming and we are trying to avoid that. Hope this makes sense.

**Comment by bigfish24** [comment]
@MattW86 that is great feedback and we hope we can support this in the future. I will say it is a lower priority feature at the moment. Right now our highest priority is custom authentication support and improving the dashboard (lots we plan to do there).
